### ISSUE TITLE
Proposal: init-declarations (using never) ignore for loops

### ISSUE THREAD

**Issue by PhilVargas** [eslint/eslint#3641]
init-declarations set to never does not allow variables to be assigned on declaration, however this doesn't mix well when using a for loop. the typical use case of this setting is like so:

```
let foo;
foo = require('foo-module');
...
```

however a common use of a for loop inits the counter variable inline
```
for(var i = 0; i < bar.length; i++){ ... }
```

this raises a warning from `init-declarations: [1, "never"]`. 

ideally, i'd like to see an option present to not raise this error in some cases. e.g. inside of forloops: 
`init-declarations: [1, "never", { "ignoreForLoops": true }`

using v1.3.1

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
Is it just in `for` loops you'd like an exception?

**Comment by PhilVargas** [comment]
its the only common deviance that comes to mind.

**Comment by ilyavolodin** [comment]
I would imagine this should apply to `while` loops as well, and `for..in, for...of` in that case.

**Comment by PhilVargas** [comment]
Ah yes. Generally, I figured anything that uses a temp variable could apply to this rule. Thanks for pointing those out!

**Comment by michaelficarra** [comment]
@ilyavolodin `while` doesn't have an initialiser.

**Comment by gyandeeps** [comment]
Are we just missing accepted label, based on the convo?
@nzakas @ilyavolodin

**Comment by ilyavolodin** [comment]
I think it's a reasonable option to add.

**Comment by nzakas** [comment]
OK, open for implementation.

**Comment by alberto** [comment]
Working on this. Should it be an option or the default behavior of `never`?

**Comment by ilyavolodin** [comment]
Proposed option is: `init-declarations: [1, "never", { "ignoreForLoops": true }]` I think that should work

**Comment by alberto** [comment]
Is `ignoreForLoops` an appropriate name considering that initializations inside the `for` body are still validated?

**Comment by ilyavolodin** [comment]
Hmm.. Good point. Not sure what would be a better name...

**Comment by nzakas** [comment]
IgnoreForLoopInit?

**Comment by ilyavolodin** [comment]
I guess that would work. Pretty abstract name though, but I can't come up with anything better.

**Comment by alberto** [comment]
Updated (camelcased for consistency with other options).

**Comment by PhilVargas** [comment]
oh sweet. Glad to see this made it :+1:
